### DISCUSSION TITLE
Process update without changing package file

### DISCUSSION THREAD

**Discussion by cmplank** [renovatebot/renovate#13653] [Request Help]
I know this is an odd request. I'm trying to use `postUpgradeTasks` to run a command that updates libraries (and other files) to a particular version. But if the version is already updated in `package.json`, the command won't do its job. So I would like some way to tell Renovate not to update the package.json, but still run the postUpgradeTask which injects the version that it is upgrading to.

Here is what I'm trying to do that has gotten me into this situation. I'm using `postUpgradeTasks` to upgrade all angular dependencies (`@angular/core, @angular/cli, @angular/cdk, and @angular/material`). I'm doing this because they are all versioned together and require as peer dependencies the same major versions of each other. Here is my config:

```js
    {
      matchPackagePrefix: ["@angular"],
      groupName: "All angular monorepos together",
      postUpgradeTasks: {
        commands: [
          // Because my package.json is in a subproject directory, I have to cd to it prior to running commands.
          // The parentheses groups the commands so that the directory isn't changed in subsequent commands after the ending parens.
          "(cd {{{packageFileDir}}} && npm install --ignore-scripts --force)",
          "(cd {{{packageFileDir}}} && npx ng update {{{depName}}} --migrate-only --from={{{currentVersion}}} --to={{{newVersion}}} --allow-dirty --force)"
        ],
        fileFilters: ["**/**"]
      },
      separateMultipleMajor: true
    },
```

The trouble is that when I match all dependencies with the `@angular` prefix, it tries to run `ng update` for dependencies that are managed by `@angular/core`, etc and generates errors like these on the PR.

```
Command failed: (cd my-project && npx ng update @angular/common --migrate-only --from=11.2.14 --to=12.2.15 --allow-dirty --force) 
Repository is not clean. Update changes will be mixed with pre-existing changes.
Package does not provide migrations.
```

But when I limit the package matching to only the projects that actually can handle `ng update` (["@angular/cli", "@angular/core", "@angular/cdk", "@angular/material"]), it doesn't update the other libraries because I'm using the `--migrate-only` flag of `ng update` which assumes that you've already updated the package.json. So the PR ends up like this:

```js
"dependencies": {
    "@angular/animations": "11.2.14",
    "@angular/cdk": "12.2.13",
    "@angular/common": "11.2.14",
    "@angular/compiler": "11.2.14",
    "@angular/core": "12.2.15",
    "@angular/forms": "11.2.14",
   ... etc
}
```

This would all be easier if I could just run `npx ng update {{{depName}}}@{{{newVersion}}} --allow-dirty --force` which would update the package.json appropriately. However, when the package.json is updated by Renovate prior to running this command, it presumes that all necessary changes have already been made.

So the solution I'm looking for is to stop Renovate from updating the package.json for this set of dependencies so that the postUpgradeTasks can handle it using the version that Renovate supplies.

Is this currently possible? If not, is there any interest in adding such a feature?

**Comment by rarkins** [comment]
Maybe your script can first revert the change using something like `sed`?

**Reply by cmplank** [reply]
I considered running `git checkout -- {{{packageFileDir}}}/package.json` as a first step, but it's a bit invasive into Renovate's processes so I was avoiding it. I could try it though.

I'm not sure how I would approach reverting the file using `sed`. Any thoughts on that or would `git checkout` be just as good from your perspective?

**Reply by rarkins** [reply]
Git could also be another way of doing it. Renovate doesn't commit files until once all updates are done.

Another possibility could be to use npm itself to install hte current version?

**Reply by cmplank** [reply]
Now that I'm trying this out, I see that I need to undo the package.json file only once, but I need to run `ng update` for each of the 4 major angular dependencies which may be on different patch versions. This means I can't switch `postUpgradeTasks.executeMode` to "branch" as I had been hoping to make this approach work.

Maybe I can only run once by adding more shell trickery like setting an env var and then short circuiting if it exists already... this is getting pretty hacky.
